elki-core-math/src/main/java/elki/math/statistics/tests/mcde/MWPTest.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * This file is part of ELKI:
+ * Environment for Developing KDD-Applications Supported by Index-Structures
+ *
+ * Copyright (C) 2019
+ * ELKI Development Team
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package elki.math.statistics.tests.mcde;
+
+import elki.utilities.datastructures.arraylike.NumberArrayAdapter;
+import elki.utilities.documentation.Reference;
+import elki.utilities.exceptions.AbortException;
+import elki.utilities.random.RandomFactory;
+
+import java.lang.reflect.Array;
+
+import static elki.math.statistics.distribution.NormalDistribution.erf;
+
+/**
+ * Implementation of Mann-Whitney U test returning the p-value (not the test
+ * statistic, thus MWP) for MCDEDependenceMeasure. Implements algorithm 1 and 3
+ * of reference paper.
+ *
+ * @author Alan Mazankiewicz
+ * @author Edouard Fouché
+ */
+
+@Reference(authors = "Edouard Fouché, Klemens Böhm", //
+    title = "Monte Carlo Density Estimation", //
+    booktitle = "Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)", url = "https://doi.org/10.1145/3335783.3335795", //
+    bibkey = "DBLP:conf/ssdbm/FoucheB19")
+
+public class MWPTest extends MCDETest<MWPTest.MwpIndex> {
+
+  private static final long two_times_sqrt_max_long = 2 * (long) Math.sqrt(Long.MAX_VALUE);
```

Comment:
Constants should be uppercase. This constant could be named by its meaning for the algorithm, not just the mathematical definition. For example `MAX_SAMPLES` or so. But I do not think this is actually needed, as explained below.